### PR TITLE
Squashed commit of the following:

### PR DESCRIPTION
**Pull request by kefeimo**
commit e06f125712c329899513f6231149d4db97756907
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Thu Oct 31 10:20:15 2024 -0500

    clean-up

commit 23e08563ee59c82a7ab7c2365127b456f45bd08d
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Tue Oct 29 20:40:01 2024 -0500

    needs to run the tests separately.

commit 58f44ef335efff60800638c05813f3157ec6d23c
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Tue Oct 29 18:00:43 2024 -0500

    flaky tests watch out

commit 5fddd62c6379a6bb263b649c08eb2b4ec7d5d533
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Tue Oct 29 17:28:56 2024 -0500

    fixed result_master = master_new.soe_handler.db

commit 5b899e112cfc53d0fd924e8bf6616e7a6ef49e3c
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Tue Oct 29 16:52:03 2024 -0500

    manually fixed connection error

commit c766646a8547488c3e99a6fc68d1611eb640eb75
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Tue Oct 29 15:23:17 2024 -0500

    cleaned up

commit 214ef8712bc3ea3cd04ee2e9065a9bc1c2981322
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Tue Oct 29 14:44:27 2024 -0500

    refactored to use station_application

commit 5c49385c4c0ec0f5c2ba3801926e78d06c9c7033
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Tue Oct 29 12:10:33 2024 -0500

    refactored to use outstation_app

commit 644c0d020f05517e8d1f96b10a16eb0f534dc736
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Mon Oct 28 16:51:50 2024 -0500

    from __future__ import annotations

commit 5a8108e7389766c2bd553b0923acaa1db1bde5cc
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Mon Oct 28 16:28:59 2024 -0500

    improved print out db as a screenshot

commit 0513cc119f73e20b38908ea3c4663d3927e88324
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Mon Oct 28 12:19:56 2024 -0500

    refactored dnp3station, testing, db size, dnp3demo

commit e28bb3446d477ab5f5c6aa47e389f2fa416b6afb
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Mon Oct 28 11:18:59 2024 -0500

    added testing outstation.

commit ca1e2d3cd658fa7b525c3c68214ae95c1f7ade46
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Thu Oct 24 14:10:03 2024 -0500

    pybind11 roll back

commit 1914a5d5f950be285fd49ec2ce5a522233a136fc
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Thu Oct 24 14:05:55 2024 -0500

    independent github workflow

commit aa57fade37dffab7cc00aa059f2e926b5fbd92bb
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Thu Oct 24 13:53:04 2024 -0500

    optimized github action workflow

commit 0ce490c319292b3992e14d99cd926340f79ed6f6
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Thu Oct 24 13:46:06 2024 -0500

    added "pip install pytest"

commit 9eb6438b8e36cad716af1ddb963b7b2917662795
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Thu Oct 24 13:38:51 2024 -0500

    rename

commit 2f54a284ea3e7366aa2710592a75c9bb092ace30
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Thu Oct 24 13:34:57 2024 -0500

    pytest -s -vv /tests/test_dnp3_python/*

commit 7c6292d5b063c98f83b3e432c068bb8b3987ec09
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Thu Oct 24 13:28:08 2024 -0500

    added

commit 6f6b2b2ef6d7ebde1e88f33f0daa9c2da3567985
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Thu Oct 24 13:26:05 2024 -0500

    added test_master

commit 3e4f0492290bf68bf68e32ef8912a77ce2890b20
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Wed Oct 23 16:44:56 2024 -0500

    added fixture testing

commit fe6699f1b20b721ecb616b33c83896f8d3bf09e3
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Wed Oct 23 13:55:18 2024 -0500

    modified install-package.yml

commit 5ce38dee51224085e629df5608e8ee491c562650
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Wed Oct 23 13:48:23 2024 -0500

    fiexed 3.10 considered as 3.1 in github action

commit 3782b4713df9cf6d2385c2ab8408307893fcf482
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Wed Oct 23 13:45:00 2024 -0500

    added Install Package Test for github action

commit 2fe0c2b0cffe4b731c8db43ba43d7283d60083b7
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Wed Oct 23 13:38:01 2024 -0500

    added test_insatllation, added __version__

commit 29590afb32848e638975716fa53154bd582372b0
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Wed Oct 23 11:21:34 2024 -0500

    rebase from v0.2.3b10

commit 1617afc6aedfb85bb46a0340fdd36be69ca066a6
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Wed Sep 18 15:33:53 2024 -0500

    adding new OutStationApplication class for beter usability

commit 1a1fda226929a10ee98c630f0863a725e08c0763
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Wed Aug 21 16:57:44 2024 -0500

    added notes to ScanAllObjects to improve readability

commit dd534a8d63b83f51f865ff5f84d06ee03baa4a27
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Wed Aug 21 14:14:35 2024 -0500

    added example for Measure(value, flags, dnp3time) in data_retrieval_demo

commit 270637c5ec393046c9d1e6f1375a464652dc0434
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Mon Aug 12 09:56:13 2024 -0500

    added notebook

commit 97780a0c80e89c7a326b7d4d042050157e901429
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Wed Aug 7 17:21:36 2024 -0500

    mapped to pnnl database schema

commit 9c6888a7643ab500a25a308dd96e35212e83a959
Author: Kefei Mo <kefeimo@gmail.com>
Date:   Wed Aug 7 13:28:29 2024 -0500

    added cmd sub-command to allow config db sizes.

### CODE REVIEW

File: src/dnp3_python/dnp3station/outstation.py [hunk 1]
```diff
@@ -550,11 +550,11 @@ class OutStationApplication:
 
     def __init__(
         self,
-        outstation_ip: str = "0.0.0.0",
-        port: int = 20000,
-        master_id: int = 2,
-        outstation_id: int = 1,
-        concurrency_hint: int = 1,
+        outstation_ip: str | None = "0.0.0.0",
```
Comment (craigpnnl) [inline src/dnp3_python/dnp3station/outstation.py]: Explain why these are necessary as you are setting the default to a constant so they will never be none?

**Review by craigpnnl — CHANGES_REQUESTED** [review]

File: src/dnp3_python/dnp3station/outstation.py (hunk 1 above)
Comment (kefeimo) [reply src/dnp3_python/dnp3station/outstation.py]: This is to pass the type hint checking issue.

Say, 

When init the OutstationApplication(outstation_ip=outstation_ip), the outstation_ip is retrieved from terminal input, which can be None. And the type hint validation will complain about this.

Another option is use outstation_ip: str | None = None, then `if outstaion_ip is None: outstaion_ip = "0.0.0.0"`, which can be more tedious.

**Comment by craigpnnl** [comment]
To pass what type hint checking thing?  I think it is much more readable to leave the str without the | None and update your type hinting to ignore the None.

All things could be None and require that that is checked, regardless of what is passed right?

**Comment by kefeimo** [comment]
> To pass what type hint checking thing? I think it is much more readable to leave the str without the | None and update your type hinting to ignore the None.
> 
> All things could be None and require that that is checked, regardless of what is passed right?

Make sense. Modified accordingly with [polished cmd parse input](https://github.com/VOLTTRON/dnp3-python/pull/53/commits/28f4f838db51e29063fb6c22a72f8b9cc68f2849)

File: src/dnp3_python/dnp3station/station_utils.py [hunk 2]
```diff
@@ -535,3 +579,252 @@ def to_flat_db(db: dict) -> dict:
             db_flat["Value"] += values
             db_flat["Type"] += types
     return db_flat
+
+
+# def to_pnnl_schema(db: dict, is_wrapped_text=True) -> pd.DataFrame:
```
Comment (craigpnnl) [inline src/dnp3_python/dnp3station/station_utils.py]: Should we remove all of this commented out code to clean this up?

**Comment by craigpnnl** [comment]
@kefeimo these tests run I assume on your system.  The tests won't run since they are new until this goes to the default (main) branch.

**Review by craigpnnl — APPROVED** [review]

**Pull request merged**